[MiscDiagnostics] Update diagnostics for IUO to Any coercion

theblixguy opened the pull request:
This PR updates the diagnostics for implicit coercion from `T!` to `Any`.
Resolves [SR-10199](https://bugs.swift.org/browse/SR-10199).

theblixguy commented:
cc @xedin

theblixguy commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -2697,6 +2697,14 @@ ConstraintSystem::matchTypes(Type type1, Type type2, ConstraintKind kind,
+    auto *fix = AllowImplicitCoercionToAny::create(
+        *this, type1, type2, getConstraintLocator(locator));
+    recordFix(fix);
+    return matchTypes(type1->lookThroughAllOptionalTypes(), type2,
```
This was giving me some trouble and still is - I am not sure if this is the right logic.

theblixguy commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -2697,6 +2697,14 @@ ConstraintSystem::matchTypes(Type type1, Type type2, ConstraintKind kind,
+    auto *fix = AllowImplicitCoercionToAny::create(
+        *this, type1, type2, getConstraintLocator(locator));
+    recordFix(fix);
+    return matchTypes(type1->lookThroughAllOptionalTypes(), type2,
```
Adding the fix to the conversionsOrFixes array doesn’t result in the fix being applied, so had to call recordFix.

Although there’s definitely an issue with this logic because it’s triggering the warning for a few cases where it shouldn’t be doing that.

xedin replied:
Since implicit coercions like these are allowed `repairFailures` wouldn't even be triggered, fix should be attached somewhere in https://github.com/apple/swift/blob/master/lib/Sema/CSSimplify.cpp#L2409L2426 . It's okay to use `recordFix` directly in this situation.

theblixguy replied:
Okay, done!

xedin commented on lib/Sema/CSFix.h:
```diff
@@ -136,6 +136,9 @@ enum class FixKind : uint8_t {
   AllowInaccessibleMember,
+
+  /// Allow implicit coersion from optional to any
+  AllowImplicitCoercionToAny,
```
I think it would make sense to rename this to `WarnImplicitCoercionToAny` or something like that because these are allowed already (same goes for the fix class).

theblixguy replied:
Done!

xedin commented on lib/Sema/CSDiagnostics.cpp:
```diff
@@ -2352,3 +2352,133 @@ bool InaccessibleMemberFailure::diagnoseAsError() {
+    }
+  }
+
+  auto hasImplicitlyUnwrappedResult = [&](Expr *E) -> bool {
```
This information has to be looked up in the constraint system by locator, subscripts and other expressions wouldn't have overload choices  set until `ExprRewritter` runs which is not going to happen when problems are diagnosed via fixes.

theblixguy replied:
Done!

xedin commented on lib/Sema/CSDiagnostics.cpp:
```diff
@@ -2352,3 +2352,133 @@ bool InaccessibleMemberFailure::diagnoseAsError() {
+    }
+  };
+
+  auto emitSilenceOptionalAnyWarningWithCoercion = [&](Expr *E, Type destType) {
```
Too bad this logic as well as `isOptionalToAnyCoercion`  has to be duplicated between MiscDiagnostics and here... It would be great to find a better place for this.

xedin commented on lib/Sema/CSDiagnostics.cpp:
```diff
@@ -2352,3 +2352,133 @@ bool InaccessibleMemberFailure::diagnoseAsError() {
+    auto name = UDE->getName();
+    auto parent = getDC()->getParentForLookup();
+    SmallVector<ValueDecl *, 4> lookupResults;
+    parent->lookupQualified(FromType->lookThroughAllOptionalTypes(), name,
```
This should be retrieved from constraint system. It holds the choice associated with given member reference, locator would be `(E, ConstraintLocator::Member)` or `(E, ConstraintLocator::ConstructorMember)` depending on that the base was.

theblixguy replied:
Done!

theblixguy commented:
Could you take another look @xedin? I ran the tests locally and have got a few failing ones, all for implicit coercions from IUO to Any. I'll fix them now

xedin commented on lib/Sema/CSDiagnostics.cpp:
```diff
@@ -2352,3 +2352,100 @@ bool InaccessibleMemberFailure::diagnoseAsError() {
+  auto &CS = getConstraintSystem();
+
+  auto locator = CS.getConstraintLocator(E, ConstraintLocator::Member);
+  auto resolvedOverload = getResolvedOverload(locator);
```
I'd suggest you use `getOverloadChoiceIfAvailable` here.

theblixguy replied:
Done!

xedin commented on lib/Sema/CSDiagnostics.h:
```diff
@@ -983,6 +983,19 @@ class InaccessibleMemberFailure final : public FailureDiagnostic {
   bool diagnoseAsError() override;
 };
 
+class ImplicitCoercionToAnyFailure final : public FailureDiagnostic {
```
`ImplicitCoercionToAnyFailure` -> `ImplicitCoercionToAnyWarning`

theblixguy replied:
Done!

xedin reviewed (commented):
Looks much better now, few more comments inline and test-case is missing.

theblixguy commented:
Thanks @xedin, I have made the changes and updated the existing tests. Not sure if I should add another test case because the existing ones are good enough?

theblixguy commented on test/Sema/diag_unintended_optional_behavior.swift:
```diff
@@ -7,76 +7,44 @@ func takeAny(_ left: Any, _ right: Any) -> Int? {
 func throwing() throws -> Int? {}
 
 func warnOptionalToAnyCoercion(value x: Int?) -> Any {
-  let a: Any = x // expected-warning {{expression implicitly coerced from 'Int?' to 'Any'}}
```
We should be emitting a warning here as `x` isn't an IUO. Let me take a look at what's going on!

xedin commented on test/Sema/diag_unintended_optional_behavior.swift:
```diff
@@ -7,76 +7,44 @@ func takeAny(_ left: Any, _ right: Any) -> Int? {
-  // expected-note@-1 {{provide a default value to avoid this warning}}
-  // expected-note@-2 {{force-unwrap the value to avoid this warning}}
-  // expected-note@-3 {{explicitly cast to 'Any' with 'as Any' to silence this warning}}
+  let _: Any = (x)
```
This should also be diagnosed since x is an `Int?`

xedin commented on test/Sema/diag_unintended_optional_behavior.swift:
```diff
@@ -7,76 +7,44 @@ func takeAny(_ left: Any, _ right: Any) -> Int? {
-  // expected-note@-1 {{provide a default value to avoid this warning}}
-  // expected-note@-2 {{force-unwrap the value to avoid this warning}}
-  // expected-note@-3 {{explicitly cast to 'Any' with 'as Any' to silence this warning}}
+  _ = takeAny(f as? Int, g)
```
Casts here and later have to be diagnosed as well because `as?` would result in an optional type.

xedin commented on test/Sema/diag_unintended_optional_behavior_swift_5.swift:
```diff
@@ -28,64 +25,23 @@ class D {
-  // expected-note@-5 {{provide a default value to avoid this warning}}{{49-49= ?? <#default value#>}}
-  // expected-note@-6 {{force-unwrap the value to avoid this warning}}{{49-49=!}}
-  // expected-note@-7 {{explicitly cast to 'Any' with 'as Any' to silence this warning}}{{49-49= as Any}}
-  _ = takeAny(C().a, C().b) // expected-warning {{expression implicitly coerced from 'Int?' to 'Any'}}
```
This one should be diagnosed because `b` is effectively `Any?`

xedin commented on lib/Sema/CSDiagnostics.cpp:
```diff
@@ -2352,3 +2352,100 @@ bool InaccessibleMemberFailure::diagnoseAsError() {
+  auto &TC = getTypeChecker();
+  auto &CS = getConstraintSystem();
+
+  auto locator = CS.getConstraintLocator(E, ConstraintLocator::Member);
```
Actually I overlooked this originally but this would be incorrect for anything which is not in form of `a.b`, calls have their own locator, as do unresolved members e.g. `.foo`, and subscripts, you'd have to check each one here to build right locator.

theblixguy replied:
Hmm what path element do we use for DeclRefExpr or ParenExpr? I couldn't find them in PathElementKind.

theblixguy replied:
I suppose if we have a DeclRefExpr, we could just get the decl directly.

xedin replied:
DeclRefExpr stores declaration yes, and ParenExpr, Tuple and others don’t have anything but type associated with them, so you’ll have to get a decl from type if it’s a nominal.

theblixguy replied:
Hmm for a SubscriptExpr, using SubscriptMember as path element gives me the subscript decl. Example:

```
 func test(_ a : CFString!) {
  let dict = NSMutableDictionary()
  let object = NSObject()
  dict[a] = object
}
```

What I want access to is `a` and not dict.

xedin replied:
I think locator would help you with that if ‘a’ was converted to Any (parameter type) - getAnchor() would point to its DeclRefExpr (used in the argument position).

theblixguy replied:
Hmm it seems like I am going the wrong way. I am checking for each kind of expression which is becoming tedious:

```
if (auto DRE = dyn_cast<DeclRefExpr>(E)) {
	/* Already handled above - we created a locator without a specific path */
} else if (auto UDE = dyn_cast<UnresolvedDotExpr>(E)) {
	locator = CS.getConstraintLocator(E, ConstraintLocator::Member);
} else if (auto SE = dyn_cast<SubscriptExpr>(E)) {
	locator = CS.getConstraintLocator(SE->getIndex()->getValueProvidingExpr());
} else if (auto CE = dyn_cast<CallExpr>(E)) {
	locator = CS.getConstraintLocator(CE->getFn());
} else if (auto MRE = dyn_cast<MemberRefExpr>(E)) {
	locator = CS.getConstraintLocator(MRE->getBase());
} else if (auto AE = dyn_cast<ApplyExpr>(E)) {
	locator = CS.getConstraintLocator(AE->getFn());
} else if (auto PE = dyn_cast<ParenExpr>(E)) {
	locator = CS.getConstraintLocator(PE->getSubExpr());
} else if (auto CCCE = dyn_cast<ConditionalCheckedCastExpr>(E)) {
	locator = CS.getConstraintLocator(CCCE->getSubExpr());
} else if (auto TE = dyn_cast<AnyTryExpr>(E)) {
	locator = CS.getConstraintLocator(TE->getSubExpr());
	  if (isa<CallExpr>(TE->getSubExpr())) {
		auto fn = dyn_cast<CallExpr>(TE->getSubExpr())->getFn();
		locator = CS.getConstraintLocator(fn);
	  }
    }
}
```

Is there a better way to get to the locator that I need to get its decl so I can check for the IUO attribute?

xedin replied:
I don't think you need to do that `getAnchor()->getSemanticsProvidingExpr()` would already point you to the right argument expression, you just need to handle `DeclRefExpr`, `UnresolvedDotExpr`/`MemberRefExpr`, `UnresolvedMemberExpr` and `OverloadedDeclRefExpr`, that's only expressions which could have declarations associated with them.

theblixguy replied:
Hmm maybe not always:

```
func warnOptionalToAnyCoercion(value x: Int?) -> Any {
  do {
    let _: Any = try throwing()
  } catch {}
}

func throwing() throws -> Int? {}
```

Here, the `getSemanticsProvidingExpr` would point to a `CallExpr`, so I guess I need to handle that. But besides that, I also need to figure out how to match these expressions to the right PathElementKind. For example, I can pass `ConstraintLocator::UnresolvedMember` for  `UnresolvedMemberExpr`, but not sure about the rest.

xedin replied:
Ah yeah and `ApplyExpr` too! For that case you'd have to look at `getFn()` expression. You'd also have to take a look at how locators are formed to generate constraints for the other cases in `CSGen.cpp`, look for `visit*Expr` methods in `ConstraintGenerator` class.

E.g. for `SubscriptExpr` it's going to be `SubscriptMember`, for `UnresolvedDotExpr` and `MemberRefExpr` it's going to be `Member` unless it's a constructor ref than it's `ConstructorMember`.

I think it would be useful to have that as a method in `FailureDiagnostic` just like`getArgumentExprFor`.

theblixguy replied:
Thanks for the suggestion @xedin! I have created a new method to return the locator for a given expression - does it look good to you?

xedin commented on lib/Sema/CSDiagnostics.cpp:
```diff
@@ -101,6 +101,52 @@ Expr *FailureDiagnostic::getArgumentExprFor(Expr *anchor) const {
+  if (isa<DeclRefExpr>(expr)) {
+    locator = CS.getConstraintLocator(expr);
+  } else if (isa<UnresolvedDeclRefExpr>(expr)) {
+    locator = CS.getConstraintLocator(expr);
```
FYI `UnresolvedDecltRefExpr` is eliminated by `PreCheckExpr` using `TypeChecker::resolveDeclRefExpr` and going to be replaced with either `DeclRefExpr`, `OverloadedDeclRefExpr` or `UnresolvedDotExpr` so there is no need to cover it here.

xedin commented on lib/Sema/CSDiagnostics.cpp:
```diff
@@ -101,6 +101,52 @@ Expr *FailureDiagnostic::getArgumentExprFor(Expr *anchor) const {
 }
 
+ConstraintLocator *
+FailureDiagnostic::getConstraintLocatorFor(Expr *expr) const {
```
Thank you! This could be improved further by making it return `Optional<SelectedOverload>` because there is a little use for locator without calling `getOverloadChoiceIfAvailable` right after, so this method would become a `FailureDiagnostic::getOverloadChoiceIfAvailable(Expr *expr)` :)

xedin commented on lib/Sema/CSDiagnostics.cpp:
```diff
@@ -101,6 +101,52 @@ Expr *FailureDiagnostic::getArgumentExprFor(Expr *anchor) const {
+    auto Fn = AE->getFn();
+    locator = CS.getConstraintLocator(Fn, ConstraintLocator::ApplyFunction);
+  } else if (isa<UnresolvedMemberExpr>(expr)) {
+    locator = CS.getConstraintLocator(expr, ConstraintLocator::MemberRefBase);
```
The type variable choice is going to be bound to has a `UnresolvedMember` because it's a result passed to `addUnresolvedMemberConstraint` in CSGen

xedin commented on lib/Sema/CSDiagnostics.cpp:
```diff
@@ -101,6 +101,52 @@ Expr *FailureDiagnostic::getArgumentExprFor(Expr *anchor) const {
+  } else if (isa<CallExpr>(expr)) {
+    auto CE = cast<CallExpr>(expr);
+    auto Fn = CE->getFn();
+    locator = CS.getConstraintLocator(Fn, ConstraintLocator::FunctionResult);
```
What you have to do in this case is to recurse into this function and give it `CE->getFn()` expression because actual choice comes from the "function expression" attached to application e.g.

```swift
struct S {
  func foo(_: Int) {}
  func foo(_: String) {}
}

func bar(_ s: S) {
  s.foo(42)
}
```

AST looks like this:

```
(call_expr type='()' location=<REPL Input>:2:5 range=[<REPL Input>:2:3 - line:2:11] arg_labels=_:
  (unresolved_dot_expr type='$T1' location=<REPL Input>:2:5 range=[<REPL Input>:2:3 - line:2:5] field 'foo' function_ref=single
    (declref_expr type='S' location=<REPL Input>:2:3 range=[<REPL Input>:2:3 - line:2:3] decl=REPL_1.(file).bar(_:).s@<REPL Input>:1:12 function_ref=unapplied))
  (paren_expr type='($T2)' location=<REPL Input>:2:9 range=[<REPL Input>:2:8 - line:2:11]
    (integer_literal_expr type='$T2' location=<REPL Input>:2:9 range=[<REPL Input>:2:9 - line:2:9] value=42 builtin_initializer=**NULL** initializer=**NULL**)))
```

and disjunction has following locator:

```
disjunction [[locator@0x7fe797801ea0 [UnresolvedDot@<REPL Input>:2:5 -> member]]]:$T1 bound to decl REPL_0.(file).S.foo@<REPL Input>:1:17 : (S) -> (Int) -> () at <REPL Input>:1:17 [[locator@0x7fe797801ea0 [UnresolvedDot@<REPL Input>:2:5 -> member]]]; or $T1 bound to decl REPL_0.(file).S.foo@<REPL Input>:2:8 : (S) -> (String) -> () at <REPL Input>:2:8 [[locator@0x7fe797801ea0 [UnresolvedDot@<REPL Input>:2:5 -> member]]];
```

xedin replied:
Note that locator is `[UnresolvedDot@<REPL Input>:2:5 -> member]`

xedin commented on lib/Sema/CSDiagnostics.cpp:
```diff
@@ -101,6 +101,52 @@ Expr *FailureDiagnostic::getArgumentExprFor(Expr *anchor) const {
+    }
+  } else if (isa<MemberRefExpr>(expr)) {
+    auto MRE = dyn_cast<MemberRefExpr>(expr);
+    if (isa<ConstructorRefCallExpr>(MRE->getBase())) {
```
Please use `getSelfForInitDelegationInConstructor(getDC(), expr)` here and in `UnresolvedDotExpr` case.

xedin reviewed (changes requested):
Still needs some work but looks much better now!

theblixguy commented:
Thank you @xedin, I have made the changes you requested. I can't use `getSelfForInitDelegationInConstructor()` for a `MemberRefExpr` because the function excepts an `UnresolvedDotExpr` as argument.

xedin commented:
@theblixguy Looks like for the `MemberRefExpr` you don't need that check at all. It always uses `Member` locator and what ever decl was associated with it - https://github.com/apple/swift/blob/master/lib/Sema/CSGen.cpp#L993

theblixguy commented:
@xedin Sounds good! However, there's one issue left now, which is that `FromType` and `ToType` are not always correct when we create the fix in CSSimplify. Consider this example:

```swift
func warnCollectionOfOptionalToAnyCoercion(_ a: [Int?], _ d: [String : Int?]) {
  takesCollectionOfAny(a, d)
}

func takesCollectionOfAny(_ a: [Any], _ d: [String : Any]) {}
```

Here, our `FromType` ends up being `Int?` rather than `[Int?]`. It's losing the array for some reason

xedin commented:
@theblixguy That's because `matchTypes` is recursive... Let me think if we could do anything to workaround. How much worse are diagnostics because of that? I think we should still point to the right place in the source at least?

theblixguy commented:
@xedin We end up saying `expression implicitly coerced from 'Int?' to 'Any'` instead of `expression implicitly coerced from '[Int?]' to '[Any]'`. We point to the right source locations, it's just the types either lose the array or one level of optionality in some other cases.

theblixguy commented:
Here's an idea from the top of my head: we can create a vector to hold a struct containing a locator and the two types. If we don't already have a struct in there with the same locator, we push one. Otherwise, we skip. Later when creating the fix, we can retrieve the struct and extract the original types and pass it to the fix.

xedin commented:
@theblixguy I'm to a fan for trying to keep a stack of types like that in matchTypes because that's on the hot path... 

Tell me one thing - in that example from your previous comment - what locator does warning have?

theblixguy commented:
@xedin The locator is 

```
locator@0x7fd1ea89bb38 [Call@/Users/suyashsrijan/Desktop/test.swift:2:3 -> apply argument -> comparing call argument #0 to parameter #0 -> generic argument #0

locator@0x7fd1ea89bad0 [Call@/Users/suyashsrijan/Desktop/test.swift:2:3 -> apply argument -> comparing call argument #1 to parameter #1 -> generic argument #1
```

xedin commented:
@theblixguy Ok, that's good! This means that if you strip away `generic argument` part of the locator and simplify it, it would point to the expression which corresponds to the argument, what to try that?

theblixguy commented:
@xedin Okay! This is only for a `CallExpr`, right? I can do `CS.getConstraintLocator(expr, ConstraintLocator::FunctionArgument)` and that gives the correct diagnostic!

xedin commented:
@theblixguy You can't actually use `CS.getConstraintLocator(expr, ConstraintLocator::FunctionArgument) ` because it's not always going to return want you think it would. I'm not actually talking about new `getOverloadChoiceIfAvailable`, that's already correct, but rather how to get `fromType` and `toType` for diagnostic message from constraint system instead of trying to save them in the fix. 

My suggestion would be to try multiple different examples and see what locators are going to be and if it would be possible to deduce types from locators (e.g. you can find type variables by locator as well by iterating over `TypeVariables` array in `ConstraintSystem`). We already figured that in case of apply expressions (and not just calls) you can get both argument and parameter based on the information stored in the locator (`ApplyArgToParam` stores both positions), I think that's the case for member references as well.

theblixguy commented:
Thanks @xedin for the suggestion! We don't always have all the type variables we need in CS. For example: 

```swift
func foo(value x: Int?) -> Any {
 let y: Any = x
}
```

in this case, we'll only have one type variable for `x`, but not for `y`. When we have at least 2 type variables, it's really simple to get the types we want - in my tests, the destination type ends up being the 1st TV and the source type ends up being the 2nd TV.

xedin commented:
Sorry I'm not say that we should away use type variables... I think that it might be okay to pinpoint types like new diagnostic would do e.g. instead of `[Int?]` mention `Int?` as long as diagnostic points to the right location in source. @jrose-apple WDYT?

theblixguy commented:
Oh sorry, I misunderstood! We can probably use a combination of type variables, locators and perhaps anchor to determine the two types (I gave it a shot and it seems to work sometimes), but at that point, I would say it starts to feel like a total hack and we might as well just store the original types in a vector in the constraint system as I originally suggested. I am not sure if there's going to be any noticeable impact on performance because we will only store it once (per locator) and it might be useful in later fixes.

Also, while we have regressed _slightly_ in terms of showing the correct type in the diagnostic, the other problem is we actually don't emit any diagnostic in some cases because the types we store in the fix aren't correct. That's something which absolutely needs to be addressed.

xedin commented:
Let’s just fix this in MiscDiagnostics and take time to experiment with warning fix.

jrose-apple commented:
I don't like losing the Array part, because there's no expression in that statement that has the type `Int?`. If it's hard to keep it intact, though, I think we could assuage my misgivings by rephrasing the text: "expression *performs* implicit coercion from 'Int?' to 'Any'".

theblixguy commented:
@xedin What if we store the type in `visitErasureExpr` in MiscDiagnostics somewhere in the type checker? We can then use those types during the fix by fetching them from the type checker?

xedin commented:
@theblixguy As far as I understand it wouldn't be required to store types if we were to fix MiscDiagnostics, right? As for the fix - I think we we'd have to refactor `matchTypes` to be able to do this properly so warnings and errors could be propagated up e.g. refactor `TypeMatchResult` to have it's own kinds e.g. `Warn, Error, Success` so we can propagate warning info up.

theblixguy commented:
@xedin Hmm yeah, should I remove the fix code for now or keep it so we can use it once we’ve fixed the underlying issue? I can’t create the fix in MiscDiagnostics because the CS isn’t setup that early. 

So I can either remove all the warning fix related code I wrote and just fix the issue in MiscDiagnostics or do that + keep the warning fix related code so we can use it once we’ve figured out how to properly pass the types from matchTypes?

xedin commented:
> So I can either remove all the warning fix related code I wrote and just fix the issue in MiscDiagnostics directly

I think that's the way to go for now.

theblixguy committed:
[diag] do not emit any diagnostics if implicitly coercing from T! to Any

theblixguy committed:
[test] update tests

theblixguy commented:
@xedin Okay, I have reverted the changes and fixed the issue directly in MiscDiagnostics! I have created a copy of the branch, so whenever `matchTypes` gets refactored, I will create a new PR to move this to the warning fix.

xedin commented:
@theblixguy Sounds great, thank you!

theblixguy commented:
Can you run the CI for me if the changes look good to you? Thank you!

theblixguy committed:
[diagnostics] check that destination type is actually Any

theblixguy commented:
Bump. Does this look good to you @xedin? Thank you!

xedin commented on test/Sema/diag_unintended_optional_behavior_swift_5.swift:
```diff
@@ -28,64 +25,23 @@ class D {
-  // expected-warning@-3 {{expression implicitly coerced from 'Any????' to 'Any?'}}
-  // expected-note@-4 {{force-unwrap the value to avoid this warning}}{{24-24=!!!}}
-  // expected-note@-5 {{explicitly cast to 'Any?' with 'as Any?' to silence this warning}}{{24-24= as Any?}}
+  takesOptionalAny(c, d)
```
Sorry for the delay, I think we should still diagnose this case because `Int??!` is effectively `Int??` which gets converted to `Any`.

theblixguy replied:
Do you mean we only allow conversions from, example, `T!` but not `T?!`, etc? In other words, if we have an ? before the !, then diagnose?

xedin replied:
Yes, that’s right.

theblixguy replied:
Thanks @xedin, I have made the changes :)

theblixguy committed:
Revert "[test] update tests"

This reverts commit e88f9d59d27b2be97727823932de4da18d32a93f.

theblixguy committed:
Merge branch 'master' into fix/SR-10199

theblixguy committed:
[diag] check for number of optionals in source type

theblixguy committed:
[test] update tests

xedin reviewed (approved):
LGTM!

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test macOS platform

hamishknight commented:
I'm a little confused why we're removing the diagnostic for an implicit coercion from an IUO to `Any`. Just like a coercion from a regular optional to `Any`, we're erasing optionality that cannot be easily recovered later, e.g:

```swift
var iuo: Int! = 0
var opt: Int? = 0

func takesAny(_ x: Any) {
  print(x)
}

takesAny(iuo) // Optional(0)
takesAny(opt) // Optional(0)
```

We could probably improve the diagnostic to point out that the IUO decl `var iuo: Int!` actually has the type `Int?`, but I don't think we should be removing the warning in this case – especially given the fact that the user might be expecting the IUO to be unwrapped on coercing to `Any`.

xedin commented:
That's a fair point, it looks like at least I need to re-read https://swift.org/blog/iuo/

theblixguy commented:
@hamishknight The diagnostic already says `Expression implicitly coerced from 'T?' to 'Any'`, so if we want to keep it in this scenario, then this PR is no-op. Are you suggesting we add a diagnostic note to `iuo` to clarify that the type is actually `T?` and not `T`?

hamishknight commented:
@theblixguy Yes, I think a note on the diagnostic would be good. Another option for the message would be something like `implicitly unwrappable %0 %1 of type %2 is not unwrapped here` (where `%0` is a `DescriptiveDeclKind`, `%1` a `DeclName` and `%2` a `Type`), e.g `implicitly unwrappable var 'iuo' of type 'Int?' is not unwrapped here`.  That might be getting a bit wordy though.

In either case, bear in mind you'll likely need to modify the message slightly for the case where you have an application of a function/subscript/constructor, as then it's not the declaration itself that's of optional type, but the return.

Also, I just noticed by inspection that when `hasImplicitlyUnwrappedResult` looks through an apply to the function decl ref, it doesn't check that it's looking at an `AbstractFunctionDecl` or `SubscriptDecl`, meaning that it would incorrectly return `true` for the following coercion:

```swift
func takesAny(_ x: Any) {}

let fn: (() -> Int?)! = { return nil }
takesAny(fn())
```

Even though `fn` has already been unwrapped before the coercion takes place. This case probably doesn't come up much in real-world usage, but would still be nice to handle correctly if you want to.

xedin commented:
@hamishknight I think this diagnostic would make current behavior even more confusing to people. IUO is supposed to be unwrapped implicitly that's the behavior as spelled in the name, current warning doesn't really make sense because we are exposing an implementation detail of the type to the users (having IUO really be an optional with a special flag), users are not supposed to know how IUO type is implemented under the hood to be able to use it, SR-10199 is a good example of how confusing this currently is.

hamishknight commented:
@xedin As I understand it, IUOs being regular optionals with a special flag on the decl for implicit unwrapping is not only the implementation, but also the mental model we want users to have for them.

This allows them to reason with code like this:

```swift
func takesInt(_ x: Int) {}

var x: Int! = 0 // `x` is of type `Int?`, but allows for unwrapping if needed
takesInt(x) // okay, we can implicitly unwrap

let y = x // `y` is of type `Int?`
takesInt(y) // error
```

as they understand that being an IUO is a feature of the decl rather than the type. It is certainly unfortunate that IUO-ness is currently spelled through the type rather than e.g a decl attribute though.

This mental model is explicitly stated in [the blog post on IUOs][iuoblog], and is also (somewhat more implicitly) stated in [the Swift Language Guide's chapter on IUOs][iuoguide] (emphasis mine):

> The Swift language defines the postfix `!` as syntactic sugar for the named type `Optional<Wrapped>`, which is defined in the Swift standard library, with the additional behavior that it’s automatically unwrapped when it’s accessed. [...] With the exception of the implicit unwrapping behavior, the following two declarations are equivalent:
> 
> ```
> var implicitlyUnwrappedString: String!
> var explicitlyUnwrappedString: Optional<String>
> ```
> [...]
> Because **implicit unwrapping changes the meaning of the declaration** that contains that type, optional types that are nested inside a tuple type or a generic type—such as the element types of a dictionary or array—can’t be marked as implicitly unwrapped.


  [iuoblog]: https://swift.org/blog/iuo/
  [iuoguide]: https://docs.swift.org/swift-book/ReferenceManual/Types.html#ID453

theblixguy commented:
Explaining to the users how an IUO type works in a diagnostic note is not great however. Currently, we're skipping this diagnostic _only_ for `T!`, but not `T?!, T??!, T??!, ...`. `T!` is effectively `T`, so a diagnostic note mentioning `T?` can be confusing. In case of `T?!` for example (which is effectively `T?`), I think it's makes sense and nothing changes in terms of diagnostics.

When I pass `T!` to a function that accepts `T` or `Any`, the mental model would be "I'm passing T" (because it's implicitly unwrapped), the fact that it is implemented as `T?` with a flag - I shouldn't have to think about that.

xedin commented:
@hamishknight The point I'm trying to get cross - IUO is supposed to be unwrapped in some conditions which are not explicitly/exhaustively specified e.g. this quote from the IUO blog post:

> In other words, you can read String! as “this value has the type Optional<String> and also carries information saying that it can be implicitly unwrapped if needed”.

or from the referenced you have quoted:

> which is defined in the Swift standard library, with the additional behavior that it’s automatically unwrapped when it’s accessed.

The reference quote makes it sound like access supposed to be unwrapped _always_ on access, which obviously doesn't much current implementation.

So the question is - is explicit unwrap needed to convert from `Int!` to `Any` not knowing any implementation details about `!` but only that it "can be implicitly unwrapped if needed”? That's a hard question and as author of SR-10199 points out it's an unexpected behavior to need to unwrap something which is already supposed to be implicitly unwrapped. 

I think if we want to leave the warning we need to make it crystal clear what/why it is happening.

hamishknight commented:
@xedin I completely agree that the documentation is a bit lacking in specifying when an IUO will be unwrapped (and indeed misleading by implying that it's always unwrapped!). It would be nice if we could update it to include some along the lines of this paragraph from [SE-0054] which I think summarises the behaviour quite nicely:

> If the expression can be explicitly type checked with a strong optional type, it will be. However, the type checker will fall back to forcing the optional if necessary. The effect of this behavior is that the result of any expression that refers to a value declared as `T!` will either have type `T` or type `T?`.

Following such a rule should lead the user to the conclusion that, given an IUO `x`, `x as Any` does not perform an unwrap, as the expression can be type-checked using `x` as a strong optional type (and this conclusion should be reinforced by the `Expression implicitly coerced from 'T?' to 'Any'` diagnostic).

> I think if we want to leave the warning we need to make it crystal clear what/why it is happening.

I agree, hence adding the note. Do you think the primary diagnostic should be changed instead?


  [SE-0054]: https://github.com/apple/swift-evolution/blob/master/proposals/0054-abolish-iuo.md

xedin commented:
Sorry for the delay, got side-tracked by other stuff.

> I agree, hence adding the note. Do you think the primary diagnostic should be changed instead?

I think it would make sense to have the diagnostic itself clearly state what is going on for the people who have not yet changed their mental model to be able to understand what is going on.

theblixguy commented:
Maybe we can say something like `Expression implicitly coerced from 'T?' to 'Any' because '<decl>' is declared as implicitly unwrappable with '!'`?

hamishknight commented:
@theblixguy IMO that could be a bit confusing due to the fact that the coercion from an optional value to `Any` occurs despite being an IUO, not because of it. How about something like `coercion of implicitly unwrappable value of type 'T?' to 'Any' does not unwrap optional`?

xedin commented:
@hamishknight Your suggestion sounds good to me!

theblixguy commented:
So basically:

`Expression implicitly coerced from 'T?' to 'Any'` -> `coercion of implicitly unwrappable value of type 'T?' to 'Any' does not unwrap optional`

and a note attached to the decl saying `implicitly unwrapped <value>/<function> declared here`?

xedin commented:
Yeah, that should do it.

theblixguy commented:
Okay, I'll update this PR with the change!

theblixguy committed:
Merge branch 'master' into fix/SR-10199

theblixguy committed:
[MiscDiagnostics] Update optional to any diagnostic message and add a new note

theblixguy committed:
Revert "[test] update tests"

This reverts commit 5456455a575759592b1fbec9c6e5a012ac619939.

theblixguy commented:
@xedin I have updated the diagnostics, does it look good to you? I'll run tests locally and update the test files after that.

xedin commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -3011,7 +3011,11 @@ WARNING(optional_pattern_match_promotion,none,
+        "unwrap optional", (Type, Type))
+NOTE(optional_to_any_coercion_note,none,
+     "implicitly unwrapped %select{variable|parameter|function}0 %1 declared "
+     "here", (unsigned, Identifier))
```
Switch from `unsigned` to `DescriptiveDeclKind` here.

xedin commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -3460,6 +3460,21 @@ static void diagnoseUnintendedOptionalBehavior(TypeChecker &TC, const Expr *E,
         .highlight(subExpr->getSourceRange());
 
+      if (auto decl = getDeclForExpr(subExpr)) {
+        auto kind = [&](ValueDecl *decl) -> unsigned {
```
Instead of doing this use `Decl::getDescriptiveKind()` in the diagnostic.

theblixguy replied:
The problem with that is the diagnostics isn't as nice - you would get, for example, `implicitly unwrapped let declared here` vs. `implicitly unwrapped variable 'fk' declared here`. Rather than using `let`, `var`, `func`, it might be better to use the full name and combine let/var into `variable`. WDTY?

xedin replied:
I think it's okay to use `let` here, I'd prefer we stick to the descriptive kind instead of trying to invent our own here. Actually we might not even need "implicitly unwrapped" bit here since it's going to be attached to an error that already has that mentioned.

hamishknight commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -3011,7 +3011,11 @@ WARNING(optional_pattern_match_promotion,none,
         (Type, Type))
 WARNING(optional_to_any_coercion,none,
-        "expression implicitly coerced from %0 to %1", (Type, Type))
+        "coercion of implicitly unwrappable value of type %0 to %1 does not "
```
This needs to be a separate diagnostic to `optional_to_any_coercion` that we diagnose specifically for IUOs, as we still want to emit `expression implicitly coerced from 'T?' to 'Any'` for plain non-IUO optional-to-Any coercions.

theblixguy replied:
Fixed.

hamishknight replied:
@theblixguy Thanks! When you get round to updating the tests, could you please add the following test case (from [my previous comment](https://github.com/apple/swift/pull/23617#issuecomment-481885817)):

```swift
let fn: (() -> Int?)! = { nil }
takeAny(fn())
```

as I believe that will incorrectly trigger the IUO diagnostic rather than the plain optional one. It's up to you whether you want to fix the logic in `hasImplicitlyUnwrappedResult` in this PR – if not, you can just leave a FIXME on it.

theblixguy replied:
Yes indeed it triggers the IUO diagnostic. Maybe we can check if the return type of a function or subscript is optional or not, besides the IUO attribute on the decl?

hamishknight replied:
@theblixguy The problem with checking the return type is that it'll be optional for both IUOs and regular optionals. How about checking that when you look through an `ApplyExpr`, that the decl you end up with is an `AbstractFunctionDecl`? e.g:

```cxx
    static bool hasImplicitlyUnwrappedResult(Expr *E) {
      auto *decl = getDeclForImplicitlyUnwrappedAttr(E);
      return decl
        && decl->getAttrs().hasAttribute<ImplicitlyUnwrappedOptionalAttr>();
    }

    static ValueDecl *getDeclForImplicitlyUnwrappedAttr(Expr *E) {
      E = E->getValueProvidingExpr();
      
      // Look through implicit conversions like loads, derived-to-base
      // conversion, etc.
      if (auto *ICE = dyn_cast<ImplicitConversionExpr>(E))
        E = ICE->getSubExpr();

      if (auto *subscript = dyn_cast<SubscriptExpr>(E)) {
        if (subscript->hasDecl())
          return subscript->getDecl().getDecl();

        return nullptr;
      }

      if (auto *memberRef = dyn_cast<MemberRefExpr>(E))
        return memberRef->getMember().getDecl();

      if (auto *declRef = dyn_cast<DeclRefExpr>(E))
        return declRef->getDecl();

      // Look through an apply to see if we're calling a function decl, which
      // might have an IUO attribute indicating that the return value is an IUO.
      if (auto *apply = dyn_cast<ApplyExpr>(E)) {
        auto *decl = apply->getCalledValue();
        if (decl && isa<AbstractFunctionDecl>(decl))
          return decl;
      }
      return nullptr;
    }
```

theblixguy committed:
[MiscDiagnostics] Only emit the diagnostic if we have an IUO result

theblixguy committed:
[Misc Diagnostics] Update getExprForDecl, use DescriptiveDeclKind, etc

theblixguy commented:
I have updated the tests with the new diagnostic - ran it locally and all of them pass. @xedin could you run the smoke tests if everything looks good to you?

xedin commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -3012,6 +3012,11 @@ WARNING(optional_pattern_match_promotion,none,
+        "coercion of implicitly unwrappable value of type %0 to %1 does not "
+        "unwrap optional", (Type, Type))
+NOTE(iuo_to_any_coercion_note,none,
+     "implicitly unwrapped %0 '%1' declared here", (DescriptiveDeclKind, StringRef))
```
I think it makes more sense to add a separate note for function declarations which is going to say `%0 '%1' with implicitly unwrapped result type is declared here`. Also you should use `DeclName` or at least `Identifier` for both notes instead of a string.

theblixguy replied:
Hmm okay! I have added a new diagnostic note for functions. I tried using `Identifier`, however it would trigger an asset related to special names. `DeclName` makes sense, so I have switched to that instead of `StringRef`.

xedin replied:
Yes, it makes sense if the name was special (e.g. constructor or subscript) and you tried to do `getBaseIdentifier` but it should work for `getBaseName()` (at least after looking at `<<` operator implementation for `DeclBaseName` which uses `userFacingName()` inside).

xedin commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -3455,15 +3458,29 @@ static void diagnoseUnintendedOptionalBehavior(TypeChecker &TC, const Expr *E,
+        if (auto decl = getDeclForImplicitlyUnwrappedExpr(subExpr)) {
+          TC.diagnose(subExpr->getStartLoc(), diag::iuo_to_any_coercion,
+                      /* from */ srcType, /* to */ destType)
+          .highlight(subExpr->getSourceRange());
```
LGTM! Could you please run `clang-format` on all of these changes, they are not currently formatted at the moment.

theblixguy replied:
I did run it but not sure why it didn't indent this properly 🤷‍♂️ Anyway, it's fixed now!

theblixguy committed:
[MiscDiagnostics] Add new diagnostic note for function result, re-indent code, etc

theblixguy committed:
[Test] Update tests

xedin commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -3455,15 +3458,32 @@ static void diagnoseUnintendedOptionalBehavior(TypeChecker &TC, const Expr *E,
+                              : diag::iuo_to_any_coercion_note;
+
+          TC.diagnose(decl->getLoc(), noteDiag, decl->getDescriptiveKind(),
+                      DeclName(decl->getBaseName()));
```
How about `decl->getFullName()` ?

theblixguy replied:
Would that include the arg name as well?

xedin replied:
Yes, that would include labels.

theblixguy replied:
Yeah, which is why I didn’t use it

xedin replied:
Why not?

theblixguy replied:
Hmm I think it would be cleaner, but after thinking more about it, I guess it doesn't hurt to add arg labels as well. I have now made the change!

theblixguy committed:
[MiscDiagnostics] Use decl->getName() for IUO to Any diagnostic

xedin commented:
@swift-ci please test

theblixguy commented:
Unrelated failure on Linux: 

> 22:55:56 The following tests FAILED:
22:55:56 	  1 - TestFoundation (ILLEGAL)
22:55:56 Errors while running CTest
22:55:56 FAILED: CMakeFiles/test.util

xedin commented:
@swift-ci please test Linux platform

xedin commented:
@swift-ci please test Linux platform

theblixguy commented:
Same unrelated failure with TestFoundation - the Swift tests passed though.

xedin commented:
@swift-ci please test Linux platform

xedin commented:
It was a tough day for Linux CI yesterday, should be resolved now :)

theblixguy commented:
Thank you @xedin! The tests have passed now 🎉 I will create a PR to cherry-pick this to the 5.1 branch.

xedin commented:
No problem! Please cherry-pick a squashed commit I have merged.
